release-notes.txt: add 2024.10 release notes (#21051)
RIOT-OS/RIOT: release-notes.txt

RIOT-2024.10 "FlexTape" - Release Notes
============================
RIOT is a multi-threading operating system which enables soft real-time
capabilities and comes with support for a range of devices that are typically
found in the Internet of Things: 8-bit and 16-bit microcontrollers as well as
light-weight 32-bit processors.

RIOT is based on the following design principles: energy-efficiency, soft
real-time capabilities, small memory footprint, modularity, and uniform API
access, independent of the underlying hardware (with partial POSIX compliance).

RIOT is developed by an international open-source community which is
independent of specific vendors (e.g. similarly to the Linux community) and is
licensed with a non-viral copyleft license (LGPLv2.1), which allows indirect
business models around the free open-source software platform provided by RIOT.

About this release
==================

> The 2024.10 release includes:

Reviewer: ```suggestion
Highlights of the 2024.10 release:
```

> This release fixes some long standing pktbuf leaks that occurred when using

Reviewer: ```suggestion
This release fixes some long-standing packet buffer leaks that occurred when using
```

`gnrc_netif_pktq` and 6LoWPAN or `netdev_new_api` (#20983, #20834).

With SUIT it is now easier to accept firmwares signed with different keys,

> e.g. a node build and signed with a development key can be configured to

Reviewer: ```suggestion
e.g., a node build and signed with a development key can be configured to
```

accept images signed with either the development key or the production key (#21019).
The public keys are now also stored on the developer's machine, so if a
encrypted private key is used, you don't need the decryption password to build

> a firmware, only for signing it.

Reviewer: Is there a PR number for this part, too?

----------------

+ Rust: Add `make cargo-command` and refactoring around it (#20828)
+ static_tests: Add test for Rust code formatting rules (#20887)
+ sys/event: add API to start periodic event without initial delay (#20911)
+ sys/event: add event_sync() (#20916)
* event/timeout: event_timeout_set() enqueues immediately if timeout is
  zero (#20870)
* examples/rust: Update CoAP example modules (#20821)
* Rust: Update to riot-wrappers 0.9 (#20829)

Networking
----------

+ drivers: Add shield_llcc68 module (#20885)
+ nib.c: add interface selection rules for static link local address
  assignment (#20784)
+ sys/event: add assertion that event has a handler (#20847)
+ sys/net/application_layer/nanocoap: add `nanocoap_sock_get_non()` (#20852)
+ tests/net: add stub test for gcoap_forward_proxy (#20909)
* gcoap: Avoid reading beyond defined input buffer (#20549)
* gnrc/ipv6/nib: don't queue packets on 6lo neighbors and drop/flush
  if… (#20834)
* gnrc_ipv6_nib: Force unspecified next hop addresses (#20371)
* gnrc_sixlowpan_frag: enable `gnrc_netif_pktq` if `netdev_new_api` is
  used (#20879)
* net/dhcpv6: Improve option handling in dhcpv6 advertise (#20801)

Packages
--------

> * makefiles/docker.inc.mk: bump riotbuild docker image to last

Reviewer 2: This looks miscategirized to me

  available (#20840)
* pkg/emlearn: bump version to 0.17.1 (#20347)
* pkg/littlefs: make use of RIOT's log module instead of pkg provided
  macros (#20894)
* pkg/lvgl: Increase default LV_MEM_SIZE for 64 bit and enable tests
  for native64 (#20846)
* pkg/nanocbor - upgrade to last version (#20813)
* pkg/wamr : updated wamr version to v2.1.1 (#20795)

Boards
------

+ boards/adafruit-grand-central-m4-express: provide arduino features (#20904)
+ boards/adafruit-metro-m4-express: initial port (#20912)
+ boards/nucleo-f439zi: add ADC support (#20814)
+ boards/nucleo-g474re: add MCU table (#20851)
+ boards/nucleo64: add pinout source (#20931)
+ boards/stm32l476g-disco: add ADC support (#20793)
+ boards/stm32l476g-disco: add PWM support (#20807)

> + examples: Add asynchronous Rust example (#20830)

Reviewer 3: Thats not board specific?

* boards/ek-lm4f120x: Change internal LED macro for C2Rust
  compatibility (#20831)
* boards/nucleo-l432kc: enable CAN support (#20845)
* CI: fix true and false positives by newer codespell version (#20899)

> * drivers/led: Allow LEDn_ON to be disabled by other modules (#20833)

Reviewer: move to drivers?

CPU
---

* core/thread: always use THREAD_CREATE_STACKTEST when DEVELHELP is
  enabled (#20450)

> * cpu/cc2538: mask length byte before checking CRC [backport 2024.10] (#21038)

Reviewer: ```suggestion
* cpu/cc2538: mask length byte before checking CRC (#21038)
```

and probably in other places, too

----------------------

+ Makefile: add .DEFAULT target to further guide new users (#20895)
* dist/tools/bootterm: bump to 0.5 (#20871)
* dist/tools/cosy: take BUILD_DIR into account (#20789)
* make: export DEVELHELP (#20889)
* makefiles/arch/native: don't be pedantic (#20873)
* makefiles/suit: make use of `SUIT_SEC_PASSWORD` optional (#20862)
* makefiles/suit: store public keys, make it easier to work with
  multiple keys (#20858)

Examples
--------

+ examples/leds_shell: add example for interactive LEDs/GPIO control (#20782)
* examples/gcoap-rust: Expose more functionality (#20838)
* examples/leds_shell: use periph_gpio_mock on native (#20796)

Testing
-------

+ tests/pkg/emlearn: add model.h to repo (#20841)
* gh-actions: bump deprecated packages (#20892)
* tests/gnrc_ipv6_nib: rtr_ltime test (#20372)

API Changes
-----------

* cpu/stm32: implement `periph_gpio_ll_switch_dir` (#20805)
* drivers/periph_gpio: let gpio_read() return bool (#20936)
* drivers/periph_gpio_ll: change API to access GPIO ports (#20639)

Uncategorized
-------------
+ .clang-format: Add default clang-format configuration (#20865)
+ CODING_CONVENTIONS.md: Add preprocessor directive formatting (#20866)

> * riot.doxyfile: Update and remove deprecated options (#20897)

Reviewer 4: This belongs to `Documentation`.

And 26 minor changes.

Deprecations
============

Removals (1)
------------

- sys/ztimer: Remove the deprecated ztimer_now64 (#20826)

Bug fixes (27)
==============

* boards/common/arduino-zero: Add Arudino SPI mapping (#20884)
* cpu/native/periph_timer: add missing -lrt to linker flags (#20816)
* cpu/sam0_common/periph_adc: add work around for errata 2.1.6
  [backport 2024.10] (#21001)
* bluetil: Ensure advertisement length does not exceed pkt len (#20881)
* boards/samd5x: GCLK source for ADC must not exceed 100 MHz (#20932)
* boards/stm32f429i-disc1: fix touch screen axis [backport 2024.10] (#21014)
* build system: simplify docker image pinning (#20877)
* core/mutex: use thread_yield_higher() in mutex_unlock() (#20890)
* core/panic: don't use LOG_ functions in panic handler (#20790)
* cpu/sam0_common/periph_gpio_ll: fix gpio_get_port() and
  gpio_ll_query_conf() [backport 2024.10] (#21008)
* dns_msg: skip RDLENGTH_LENGTH field when skipping record (#20857)
* drivers/at86rf215: return ENETDOWN when interface is down [backport
  2024.10] (#21037)
* examples/lorawan: drop crazy STM32 hack (#20883)
* gcoap/forward_proxy: handle timeout case (#20915)

> * gh-actions: bump upload-artifact to v4 (#20875)

Reviewer 4: Why is this here but `gh-actions: bump deprecated packages (#20892)` is under `Testing`?

RIOT-2024.10 "FlexTape" - Release Notes
============================
RIOT is a multi-threading operating system which enables soft real-time
capabilities and comes with support for a range of devices that are typically
found in the Internet of Things: 8-bit and 16-bit microcontrollers as well as
light-weight 32-bit processors.

RIOT is based on the following design principles: energy-efficiency, soft
real-time capabilities, small memory footprint, modularity, and uniform API
access, independent of the underlying hardware (with partial POSIX compliance).

RIOT is developed by an international open-source community which is
independent of specific vendors (e.g. similarly to the Linux community) and is
licensed with a non-viral copyleft license (LGPLv2.1), which allows indirect
business models around the free open-source software platform provided by RIOT.

About this release
==================

Highlights of the 2024.10 release:

This release fixes some long-standing packet buffer leaks that occurred when using
`gnrc_netif_pktq` and 6LoWPAN or `netdev_new_api` (#20983, #20834).

> A long standing-bug in the CC2538 radio driver has been fixed when processing 
> corrupted frames, that before could render the device unresponsive. This can be 
> triggered due to radio interference, but we also believe that an attacker within range 
> of the radio could send crafted frames to reliably trigger the issue. We therefore 
> consider this a denial of service vulnerability and advise all users of the CC2538 to 
> upgrade to 2024.10 or cherry-pick 1b61216da734fcd40ae0d832b800e4e67975bdc8. More 
> details are published in the [security advisory](https://github.com/RIOT-OS/RIOT/security/advisories/GHSA-m75q-8vj8-wppw).

Reviewer 2: ```suggestion
A long standing-bug in the CC2538 radio driver has been fixed when processing
corrupted frames, that before could render the device unresponsive. This can be
triggered due to radio interference, but we also believe that an attacker within range
of the radio could send crafted frames to reliably trigger the issue. We therefore
consider this a denial of service vulnerability and advise all users of the CC2538 to
upgrade to 2024.10 or cherry-pick 1b61216da734fcd40ae0d832b800e4e67975bdc8. More
details are published in the [security advisory](https://github.com/RIOT-OS/RIOT/security/advisories/GHSA-m75q-8vj8-wppw).
```
